### PR TITLE
Convert Shared Preferences Android to Pigeon

### PR DESCRIPTION
**Pull request by AmanNegi**
## Changes

Converted the ***Android*** Implementation of shared_preferences to [Pigeon](https://pub.dev/packages/pigeon). The files `messages.g.dart` and `Messages.java` are autogenerated by Pigeon according to the template provided in `pigeons/messages.dart`. 

- I took reference from the `path_provider` package as it is already converted to pigeon.
- The tests were heavily inspired by the SharedPreferences for MacOS as it had been already converted using Pigeon.
- All the tests are running successfully and the example is also working as expected.

## Issue
Part of https://github.com/flutter/flutter/issues/117914

## Pre-launch Checklist

- [✅] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [✅] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [✅] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [✅] I signed the [CLA].
- [✅] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [✅] I listed at least one issue that this PR fixes in the description above.
- [✅] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [✅] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [✅] I updated/added relevant documentation (doc comments with `///`).
- [✅] I added new tests to check the change I am making, or this PR is [test-exempt].
- [✅] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by AmanNegi** [comment]
@tarrinneal Could you please guide me with how can I pass through the CI/CD Pipeline, the tests were running and the example is also running fine as expected.

File: packages/shared_preferences/shared_preferences_android/AUTHORS
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_android/AUTHORS:66]: Please don't make whitespace-only changes to files (especially not to remove newlines).

File: packages/shared_preferences/shared_preferences_android/CHANGELOG.md
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_android/CHANGELOG.md:3]: Converts

(See CHANGELOG style guide linked from the checklist.)

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java:4]: Please don't move all of the existing code to a different file in this PR; it makes it extremely difficult to review the actual changes. The code that was in this file should stay there for now, with just the minimal changes to adapt it to Pigeon.

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java:28]: This PR should include filling in `SharedPreferencesTest.java` with tests of the new public API being added here.

File: packages/shared_preferences/shared_preferences_android/pigeons/messages.dart
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_android/pigeons/messages.dart:20]: Why are all of these return values nullable? What does a null return indicate that is different from `true` or `false`?

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java:122]: Why are you catching exceptions everywhere just to rethrow them without changes?

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
Thanks for the contribution!

File: packages/shared_preferences/shared_preferences_android/pigeons/messages.dart
Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/pigeons/messages.dart:20]: Yes, I have kept them nullable in case some error or exception occurs, so we have the safety mechanism to handle that just in case. If you want me to change it, to non-nullable I can do that as well.

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java
Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java:4]: Okay, I will do that, sorry it's my first time. I will try to make the reviewing process as smooth as possible.

File: packages/shared_preferences/shared_preferences_android/pigeons/messages.dart
Comment (stuartmorgan) [reply packages/shared_preferences/shared_preferences_android/pigeons/messages.dart:20]: > Yes, I have kept them nullable in case some error or exception occurs, so we have the safety mechanism to handle that just in case.

I don't follow, since exceptions don't return a value. Can you describe a scenario in which null is returned?

Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/pigeons/messages.dart:20]: Oh yeah, if it succeeds then it's true otherwise it is false for all other cases. In case of exception, the execution is stopped so there is no return in the first place. That makes sense, I will change it to non-nullable

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java
Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java:122]: Sure I will ensure that I don't rethrow it.

**Comment by tarrinneal** [comment]
> @tarrinneal Could you please guide me with how can I pass through the CI/CD Pipeline, the tests were running and the example is also running fine as expected.

Seems like you got it, nice work. I'll go ahead and review this a bit later today. Thank you for contributing!

**Comment by AmanNegi** [comment]
I have made the changes but I am facing issues with File Formatting, however I'm trying to fix it. Once I fix it I will raise a review check. Sorry for the delay

**Comment by AmanNegi** [comment]
Hey, there I successfully fixed the error regarding Code Formatting.

File: packages/shared_preferences/shared_preferences_android/AUTHORS
Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/AUTHORS]: Done! :)

File: packages/shared_preferences/shared_preferences_android/CHANGELOG.md
Comment (tarrinneal) [inline packages/shared_preferences/shared_preferences_android/CHANGELOG.md:3]: Seems like your linter may have made some unwanted changes to this file

Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/CHANGELOG.md:3]: Really, sorry I didn't see that. To confirm we are on the same page, you talking about the `*` changing to `-`  right?

**Comment by AmanNegi** [comment]
Hey, there I reverted the unwanted changes caused by linting.

**Review by AmanNegi — COMMENTED** [review]
Hey there I have fixed all the issues and the reviews.

**Review by AmanNegi — COMMENTED** [review]
Fixed all reviews successfully :)

**Comment by AmanNegi** [comment]
Hey, is there some issue with the PR? Can I help with it?

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java
Comment (tarrinneal) [inline packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java:24]: If you would still like to make this change, I think another pr that does only this would be a good idea.

**Comment by stuartmorgan** [comment]
> Hey, is there some issue with the PR?

What issue are you referring to?

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java
Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java:24]: Should I change the name perhaps?

**Comment by AmanNegi** [comment]
> > Hey, is there some issue with the PR?
> 
> What issue are you referring to?

I thought it wasn't getting merged perhaps I am missing something at my end.

**Comment by stuartmorgan** [comment]
> I thought it wasn't getting merged perhaps I am missing something at my end.

PRs don't get merged before they are approved by reviewers; please read https://github.com/flutter/flutter/wiki/Tree-hygiene#getting-a-code-review (and 11 hours since your last update is not a reasonable window of time before pinging reviewers; re-review can take a while depending on people's schedules).

**Comment by AmanNegi** [comment]
I am really sorry, I will make sure not to tag any contributor/maintainer.

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java
Comment (tarrinneal) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java:24]: I think both in a follow up pr would be ideal

Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java:24]: Yeah sure, do I have to do it after this PR is merged? I am sorry for the lame question😔, this is my first time contributing.

Comment (tarrinneal) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java:24]: you can put the pr up whenever, I think it would be easier to wait till this one is landed though

Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java:24]: Okay sure! Thanks a lot again(🙂).

**Comment by AmanNegi** [comment]
Hey there, any updates regarding this PR? 😊

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
Apparently I hadn't published my review comments; they were still in Pending mode.

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java
Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java:126]: Sorry for that, I have now fixed it. Now I have mitigated the functions to handle `IOException` and the wrapper functions then convert it to RuntimeException.

File: packages/shared_preferences/shared_preferences_android/test/shared_preferences_android_test.dart
Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/test/shared_preferences_android_test.dart:14]: Okay, I will rename it. :)

File: packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java
Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java:133]: I wrote these tests so that the user can easily understand what all functions are available and what all values can be passed. I am using Fake Tests because we can't access the `SharedPreferencesPlugin` in Unit Tests. If I am wrong about it, please let me know.

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java
Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/MethodCallHandlerImpl.java:148]: Yes, sorry for removing the cleanup process. I have updated the `finally` block in all required scenarios.

File: packages/shared_preferences/shared_preferences_android/lib/shared_preferences_android.dart
Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/lib/shared_preferences_android.dart:30]: Earlier set`DataType` was used, as at the `Android` end we used to get its name from `callMethod`. However, it makes more sense to filter that at the Dart End. Either way, this function has to be created at either Dart or Android end (which maps `setDataType` string to a function according to the dataType).

Comment (stuartmorgan) [reply packages/shared_preferences/shared_preferences_android/lib/shared_preferences_android.dart:30]: I'm not asking why the code that's currently in the helper exists, I'm asking why it's in a helper that's only called from one place, where that place ends up being just a single line calling that helper, rather than that code being here. I don't see what value the indirection is providing.

File: packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java
Comment (stuartmorgan) [reply packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java:133]: > I wrote these tests so that the user can easily understand what all functions are available and what all values can be passed.

The purpose of unit tests is to test code; documentation belongs in comments.

> I am using Fake Tests because we can't access the `SharedPreferencesPlugin` in Unit Tests.

Why wouldn't unit tests be able to access code under test?

Comment (stuartmorgan) [reply packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java:133]: > Because under `SharedPreferencesPlugin.java` we are using `android.content.SharedPreferences` and other libraries which can only execute when we test on an Emulator/device.

Internal implementation details that aren't available in unit tests can be mocked out, per my comment above.

> > I took inspiration for tests from [here](https://github.com/flutter/plugins/blob/main/packages/path_provider/path_provider_android/android/src/test/java/io/flutter/plugins/pathprovider/StorageDirectoryMapperTest.java).
> 
> Let me know if I'm wrong here.

That is testing a small utility, not the plugin itself, so isn't going to be very useful as an example. But even then, that's testing actual code, not testing test mocks, so is serving a useful test purpose.

Examples of substantive unit tests include [this](https://github.com/flutter/plugins/blob/main/packages/local_auth/local_auth_android/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java) and [this](https://github.com/flutter/plugins/blob/main/packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/MethodCallHandlerImplTest.java).

Comment (AmanNegi) [reply packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java:133]: Hey there, I went through the examples and I was able to create some tests using `MockedSharedPreferencesPlugin`. 

- I tried, mocking `android.content.SharedPreferences`: I tried this method but it results in a chain of mapping methods similar to the approach I have followed.
- I don't think we will be able to test the production code like the code under `MethodCallHandlerImpl` because it requires several mocked objects with chained methods `Mockito.when()` to map methods.

[Here](https://pastebin.com/RUa5r4yg), is the updated test. I am really unsure about how to write tests for this because there are not many references of tests with packages using `Pigeon`.

**Comment by stuartmorgan-g** [comment]
We've just completed the migration of the plugin code to the flutter/packages repository, as described in https://flutter.dev/go/flutter-plugins-repo-migration, and this repository is now being archived. Unfortunately that means that all in-progress PRs here must be moved to flutter/packages.

Please see [our instructions](https://github.com/flutter/flutter/wiki/Migrating-Plugins-repository-PRs-to-Packages) for an explanation of how to move your PR, and if you have any issues moving your PR please don't hesitate to reach out in the #hackers-ecosystem channel in [Discord](https://github.com/flutter/flutter/wiki/Chat).

Our apologies that your PR was caught in this one-time transition. We're aware that it's disruptive in the short term, and appreciate your help in getting us to a better long-term state!

**Pull request closed without merging**
